### PR TITLE
[WIP] Package benchmarks

### LINKED ISSUE #297: Integrate PkgBenchmark

### PR DESCRIPTION
**Pull request by Libbum**
WIP. Have just added some space building for now.

@Datseris: Do you usually invoke this manually? Is there some github trigger/action we could use in-conjunction to alert us that a new PR will alter performance?

Closes #297

### CODE REVIEW

**Comment by Libbum** [comment]
[This discourse post](https://discourse.julialang.org/t/pkgbenchmark-jl-workflow/24472/) answered that question essentially. There are a few integrations we can do with Travis.

**Comment by Datseris** [comment]
For the billiards we run it manually yes. Perhaps for now it is not worth the time investment to automate it and we can do it after the paper to prioritize things properly.

**Comment by Libbum** [comment]
Agreed. But when we do, I just had a look over Convex.jl's repo. They've got it sorted pretty well.

**Comment by Libbum** [comment]
@Datseris can you check the latest here and see if it's at least the start of what you envision for this? I can condense things into loops etc later, so don't worry about that.

For now, I've done building a space, building a model, adding agents.

**Comment by Datseris** [comment]
Why are there so many agent types :P Can't we live with just 2 for each "type"? One with only id and pos, and one with id, pos, and two properties, one boolean and one float? We should simplify these benchmarks as much as possible.

**Comment by Libbum** [comment]
Just wanted to have a large union set since we have scale issues. They are all essentially the same.

Other than that though, is this on the right track?

**Comment by Datseris** [comment]
Here is the quick benchmark file I've written for my implementation of array space:
```julia

(model, agent_step!, model_step!) = forest_fire_array()
a = random_agent(model)
sleep(1e-9)

println("Move agent")
@btime move_agent!($a, $model);

println("node neighbors")
@btime node_neighbors($a.pos, $model);

println("Space neighbors")
@btime space_neighbors($a, $model);

function iterate_over_neighbors(a, model)
    s = 0
    for x in space_neighbors(a, model)
        s += x
    end
    return s
end

# TODO: Need to check this for veeery large set of neighbors
println("Iterate over space neighbors")
@btime iterate_over_neighbors($a, $model);
println("Iterate over position space neighbors")
@btime iterate_over_neighbors($a.pos, $model);

println("Add agent")
@btime add_agent!($model, true)

println("node neighbors")
@btime node_neighbors($a.pos, $model);
```
I've noticed that it is important to check the following:

- time to call `node_neighbors`
- time to call `agent_neighbors` with agent first argument
- time to call `agent_neigbhors` with position first argument
- time to actually do meanigful operation over these results (my new array space has non-allocating version of space neighbors and thus simply calling space neighbors is instant)

**Comment by Libbum** [comment]
Latest gets us all add and move functions covered for all spaces. 

One issue that I couldn't sort was the unique problem from #226 poping up in continuous space when trying to invoke a model with `$continuous_model`, so I've had to add the creation of the model inside the `add_agent!` tests, which is less than ideal. Can anyone see a way around that?

I'm teaching for the next few hours, but can add the neighbor tests later tonight.

**Comment by kavir1698** [comment]
Which line is the problem?

**Comment by Libbum** [comment]
[add continuous space](https://github.com/JuliaDynamics/Agents.jl/pull/299/files#diff-a13505e4c82bd6d1cd5edf85ef2bf072R170-R215)

**Comment by kavir1698** [comment]
Can you just use `deepcopy(continuousSpace)` when you want to use it in a model?
I think #226 is a feature rather than a problem.

**Comment by Libbum** [comment]
```julia
continuous_model = ABM(ContinuousAgent, ContinuousSpace(3; extend = (10.0, 10.0, 10.0)))
SUITE["continuous"]["add"]["agent"] = @benchmarkable add_agent!($continuous_agent, deepcopy($continuous_model))
```
signal (11): Segmentation fault

Doesn't look like it :/

**Comment by kavir1698** [comment]
Hmm. Running `@benchmarkable add_agent!($continuous_agent, deepcopy($continuous_model))` does not give me segfaults.

**Comment by Libbum** [comment]
When you're running `benchmarkpkg(Agents)`?

**Comment by kavir1698** [comment]
I tried it in isolation and it worked. I am now trying to run it along everything else.

**Comment by kavir1698** [comment]
Haha it broke every running program on my computer

**Comment by Libbum** [comment]
Wow. So perhaps not a useful choice? :laughing:

**Comment by kavir1698** [comment]
Well, not being able to add an agent with the same ID as an existing one is reasonable. Not being able to use the same space object in multiple models is not a big deal. It is probably only needed for situations like our benchmarking.

**Comment by Libbum** [comment]
The models that we add are relatively small, so in the end I don't think the current working implementation will hurt too much. We will just need to be aware that if we improve the continuous space model initialisation time, it will in turn alter the timing for `add_agent!` benchmarks.

**Comment by kavir1698** [comment]
Alternatively, you can use the same space/model, but add agents with different IDs.

**Comment by Libbum** [comment]
Sure, but how do you put that in a benchmark in a sane manner? You just shift the implementation complexity from the model to the agent.

**Comment by kavir1698** [comment]
> Sure, but how do you put that in a benchmark in a sane manner? You just shift the implementation complexity from the model to the agent.

I'd say instantiating a model is more demanding that starting a new agent. So the test will be more accurate.

**Comment by Libbum** [comment]
Yes, but that's not possible.

If I have `c_agent = Continuous(1, (1.0, 2.2), (0.5, 0.6))`
and I want to benchmark `@benchmarkable add_agent!($c_agent, $c_model)`

We already know that `$c_model` doesn't work. We cannot use the same model.
Additionally, even if we could, we can't change `$c_agent` during the benchmarkable call.

**Comment by Libbum** [comment]
Latest puts us to what I think is a decent enough baseline to consider.

Certainly not exhaustive&mdash;we don't do all types of nearest neighbour checks, don't have things like `sample!` in, etc.

We do have adds and moves, node and agent positions, nearest neigbours, interacting pairs as well as agent data collection (I figure the agent collector is more complex than the model one and this way there's no requirement to actually `run!` a model in the benchmark.

Can you give this a review and add comments? I think if there's nothing major we should merge this baseline and build off of it as we finish off these other changes.

**Comment by kavir1698** [comment]
> If I have `c_agent = Continuous(1, (1.0, 2.2), (0.5, 0.6))`
> and I want to benchmark `@benchmarkable add_agent!($c_agent, $c_model)`

This is what I was suggesting: `@benchmarkable add_agent!(Continuous(1, (1.0, 2.2), (0.5, 0.6)), $c_model)` and for more tests, we change the id of the agent.

**Comment by Libbum** [comment]
I'm confused, 
> we use the same model but new agents.

> and for more tests, we change the id of the agent.

That's two different things. Regardless: what's in the latest commit is functional. If you have a direct suggestion to improve it, I'm all ears, but I don't think we can do much better at present.

**Comment by kavir1698** [comment]
I cannot run the benchmarks because it runs out of memory, but I think the previous commit should work.

**Comment by Libbum** [comment]
I'm not sure you've been following my point, since no, this doesn't work. You cannot put $continuous_model in like this. 

Memory-wise. This is something I didn't consider at all, but yeah&mdash;It hits 52GB at peak. The reason for this is the `add_agent_single!` runs will spit warnings once the space gets full & I wanted to test their function rather than them erroring.

If we're running offline, then this isn't a problem, but if we want to put these on travis, then we may need to drop those particular tests and reconsider sizing.

**Comment by kavir1698** [comment]
No, I don't get why you say this doesn't work because #226 was about not being able adding agents with an existing model.

The memory problem was there before my changes.

**Comment by Libbum** [comment]
```julia
(5/5) benchmarking "continuous"...
  (1/5) benchmarking "add"...
    (1/4) benchmarking "create"...
    done (took 7.304266697 seconds)
    (2/4) benchmarking "create_pos"...
    done (took 4.954738765 seconds)
    (3/4) benchmarking "agent"...
Benchmarking 100%|█████████████████████████████████████████████████████████| Time: 0:04:19
ERROR: SQLite.SQLiteException("UNIQUE constraint failed: tab.id")
Stacktrace:
 [1] execute(::SQLite.Stmt, ::Tuple{Float64,Float64,Float64,Int64}) at /home/tim/.julia/packages/SQLite/LjMUn/src/SQLite.jl:320
 [2] execute(::SQLite.Stmt, ::Tuple{Float64,Float64,Float64,Int64}) at /home/tim/.julia/packages/SQLite/LjMUn/src/tables.jl:94
 [3] add_agent_pos! at /mnt/turtle/repos/Agents.jl/src/core/continuous_space.jl:131 [inlined]
 [4] add_agent!(::ContinuousAgent, ::AgentBasedModel{ContinuousAgent,ContinuousSpace{typeof(Agents.defvel),Tuple{Float64,Float64,Float64}},typeof(fastest),Nothing}) at /mnt/turtle/repos/Agents.jl/src/core/continuous_space.jl:144
```

Because the benchmarks fail if you do.

Yep, I'm aware of the memory isssue now. The reason for it I posted just then.

**Comment by kavir1698** [comment]
Oh I understand now. And there is no "setup".

**Comment by Libbum** [comment]
I tried it with a setup clause initially, but perhaps I didn't do it right. I'll take another shot at that.

**Comment by Libbum** [comment]
YES! That worked.

OK, so other than that, were there any other things you could see / wanted to add @kavir1698?

@Datseris take a look when you get the chance. 

Still on the cards: we need to figure out it the high memory requirement is a problem or not.

**Comment by kavir1698** [comment]
Cool. No more comments from me.

**Comment by Datseris** [comment]
Lines like this: `@benchmarkable space_neighbors(rand(1:1000), $graph_model, 5)` should have a setup part with `x = rand(1:1000)` so that generating random numbers isn't included in the benchmark timing.
```
julia> @btime rand(1:1000)
  12.413 ns (0 allocations: 0 bytes)
```
and is thus comparable with e.g. moving an agent or fidnding nearest neighbors.

---

At one of the initial posts I've added:
```julia
function iterate_over_neighbors(a, model)
    s = 0
    for x in space_neighbors(a, model)
        s += x
    end
    return s
end

# TODO: Need to check this for veeery large set of neighbors
println("Iterate over space neighbors")
@btime iterate_over_neighbors($a, $model);
println("Iterate over position space neighbors")
@btime iterate_over_neighbors($a.pos, $model);
```
please include these two benchmarks for all spaces. It is important to not test only how much time it takes to generate/find the neighbors, but also to actually iterate over them. The reason is because my new `ArraySpace` has an almost 0 cost of finding the agents because many things are pre-initialized, and an iterator is created.

---

There is one more thing, which I am not sure how to address. For my new `ArraySpace`, running `space_neighbors` with a specific `r` for the first time will take about 200-300 nanoseconds. However, any subsequent call to space neighbors will require 10 nanoseconds, because inside the `hoods` field of the space, something has been pre-initialized. If we only benchmark the first call to `space_neighbors`, this wrongly represents results. How to take this into account...?

---

What is the output of running these benchmarks...?

**Comment by Datseris** [comment]
Ummm, I now realize that having `rand` everywhere is not a good idea. A benchmarking code should be as deterministic as possible. If due to random numbers one benchmark has 1000 neighbors and the other has 10, it won't take the same amount of time to find all of them, right?

To be precise, in each benchmark the **amount of numerical operations should be deterministic**. So it is okay if the agents are random with random positions, if you can e.g. can guaranteed that all agents will have exactly 100 neighbors.

**Comment by Libbum** [comment]
I thought the `rand` calls may simulate the iteration (although granted, they should have been in a `setup` clause), but sure&mdash;I'll swap those out now.

For the `space_neighbors` thing, would this work?
`@benchmarkable = space_neighbors($a, $model) setup=(space_neighbors(a, model))`

Will paste an output once I've done those changes.

**Comment by Libbum** [comment]
Results of adbc75b
```julia
Benchmarkresults:
    Package: /mnt/turtle/repos/Agents.jl
    Date: 18 Sep 2020 - 10:29
    Package commit: dirty
    Julia commit: 96786e
    BenchmarkGroup:
        5-element BenchmarkTools.BenchmarkGroup:
          tags: ["Agents"]
          "space_creation" => 3-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graph", "grid", "continuous"]
                  "grid" => Trial(63.875 ms)
                  "graph" => Trial(2.349 ms)
                  "continuous" => Trial(121.415 μs)
          "model" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["initialise", "initialise_union"]
                  "initialise_union" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["graph", "grid", "continuous"]
                          "grid" => Trial(5.947 μs)
                          "graph" => Trial(5.641 μs)
                          "continuous" => Trial(6.230 μs)
                  "initialise" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["graph", "grid", "continuous"]
                          "grid" => Trial(1.033 μs)
                          "graph" => Trial(1.040 μs)
                          "continuous" => Trial(1.501 μs)
          "grid" => 6-element BenchmarkTools.BenchmarkGroup:
                  tags: ["add", "add_union", "move", "neighbors", "node"]
                  "add" => 7-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single", "create", "create_pos", "create_single", "create_fill"]
                          "create_single" => Trial(10.653 ms)
                          "create" => Trial(1.927 μs)
                          "agent_single" => Trial(10.636 ms)
                          "create_fill" => Trial(159.528 μs)
                          "create_pos" => Trial(216.012 μs)
                          "agent" => Trial(223.323 ns)
                          "agent_pos" => Trial(23.013 ns)
                  "add_union" => 4-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single", "agent_fill"]
                          "agent_single" => Trial(10.647 ms)
                          "agent_fill" => Trial(116.910 μs)
                          "agent" => Trial(215.533 ns)
                          "agent_pos" => Trial(37.411 ns)
                  "collect" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["init_agent", "store_agent"]
                          "store_agent" => Trial(1.277 ms)
                          "init_agent" => Trial(1.833 μs)
                  "neighbors" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["space_pos", "space_agent", "space_pos_iterate", "space_agent_iterate", "node_pos", "node_agent"]
                          "space_agent_iterate" => Trial(5.175 ms)
                          "node_pos" => Trial(153.813 ns)
                          "node_agent" => Trial(170.000 ns)
                          "space_pos" => Trial(51.473 μs)
                          "space_pos_iterate" => Trial(11.409 ms)
                          "space_agent" => Trial(30.443 μs)
                  "node" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["contents", "agents"]
                          "agents" => Trial(60.000 ns)
                          "contents" => Trial(2.305 ns)
                  "move" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["random", "pos", "single"]
                          "random" => Trial(86.164 ns)
                          "pos" => Trial(26.602 ns)
                          "single" => Trial(71.778 μs)
          "graph" => 6-element BenchmarkTools.BenchmarkGroup:
                  tags: ["add", "add_union", "move", "neighbors", "node"]
                  "add" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single", "create", "create_pos", "create_single"]
                          "create_single" => Trial(310.871 μs)
                          "create" => Trial(51.682 μs)
                          "agent_single" => Trial(330.170 μs)
                          "create_pos" => Trial(161.752 μs)
                          "agent" => Trial(96.415 ns)
                          "agent_pos" => Trial(22.107 ns)
                  "add_union" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single"]
                          "agent_single" => Trial(291.895 μs)
                          "agent" => Trial(104.836 ns)
                          "agent_pos" => Trial(35.904 ns)
                  "collect" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["init_agent", "store_agent"]
                          "store_agent" => Trial(221.623 μs)
                          "init_agent" => Trial(2.590 μs)
                  "neighbors" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["space_pos", "space_agent", "space_pos_iterate", "space_agent_iterate", "node_pos", "node_agent"]
                          "space_agent_iterate" => Trial(231.865 μs)
                          "node_pos" => Trial(25.781 ns)
                          "node_agent" => Trial(24.500 ns)
                          "space_pos" => Trial(29.637 μs)
                          "space_pos_iterate" => Trial(230.796 μs)
                          "space_agent" => Trial(30.589 μs)
                  "node" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["contents", "agents"]
                          "agents" => Trial(68.800 ns)
                          "contents" => Trial(1.727 ns)
                  "move" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["random", "pos", "single"]
                          "random" => Trial(57.871 ns)
                          "pos" => Trial(20.494 ns)
                          "single" => Trial(9.250 μs)
          "continuous" => 5-element BenchmarkTools.BenchmarkGroup:
                  tags: ["add", "add_union", "move", "neighbors"]
                  "add" => 4-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "create", "create_pos"]
                          "create" => Trial(14.705 μs)
                          "create_pos" => Trial(227.968 μs)
                          "agent" => Trial(8.166 μs)
                          "agent_pos" => Trial(5.552 μs)
                  "add_union" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos"]
                          "agent" => Trial(8.455 μs)
                          "agent_pos" => Trial(6.151 μs)
                  "collect" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["init_agent", "store_agent"]
                          "store_agent" => Trial(7.768 ms)
                          "init_agent" => Trial(2.240 μs)
                  "neighbors" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["space_pos", "space_agent", "space_pos_iterate", "space_agent_iterate", "nearest", "interacting"]
                          "space_agent_iterate" => Trial(16.126 ms)
                          "nearest" => Trial(9.899 ms)
                          "space_pos" => Trial(11.886 ms)
                          "space_pos_iterate" => Trial(15.782 ms)
                          "interacting" => Trial(149.754 s)
                          "space_agent" => Trial(9.561 ms)
                  "move" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["update", "vel"]
                          "update" => Trial(2.062 μs)
                          "vel" => Trial(2.004 μs)
```

- The `setup` fix from the continuous space model problem works so long as you don't `retune`. It will fail when tuning. I don't know if that's a super large issue or not.
- Adding a `setup` step to `space_neighbors` will be trivial but is not done yet, since I'm not sure it solves the problem

**Comment by Datseris** [comment]
> For the space_neighbors thing, would this work?
`@benchmarkable = space_neighbors($a, $model) setup=(space_neighbors(a, model))`

yes this should be good enough!

**Comment by Datseris** [comment]
The benchmarks of continuous space are abysmal... @kavir1698 when you added the code based on this SQLdata base you claimed it was "fast enough" but that is not what I see here. Moving an agent shouldn't take more than nanoseconds, so all the time is spent in updating the database. @kavir1698 I think you should do this new compartment space once I completely replace `GridSpace` with my `ArraySpace`.

File: benchmark/benchmarks.jl [hunk 1]
```diff
@@ -118,28 +138,30 @@ SUITE["graph"]["add_union"]["agent_pos"] =
...
     @benchmarkable add_agent_single!($graph_agent, $graph_union_model)
 
-graph_model = ABM(GraphAgent, GraphSpace(complete_digraph(1000)))
-for _ in 1:800
+graph_model = ABM(GraphAgent, GraphSpace(complete_digraph(2000)))
+for _ in 1:1800
     add_agent!(graph_model, 6.5, false)
 end
 a = random_agent(graph_model)
+pos = rand(1:2000)
```
Comment (Datseris) [inline benchmark/benchmarks.jl]: but this still adds agents to random positions. I would claim that we should make a complete digraph with 100 nodes and add 10 agents in each node for example.

Comment (Datseris) [reply benchmark/benchmarks.jl]: Then it would not matter from which `pos` you search, because all will have the same **amount of numerical operations**

**Comment by Datseris** [comment]
for the benchmaks I think you should initialize grid space so that every node has let's say 4 agents.

**Comment by kavir1698** [comment]
> The benchmarks of continuous space are abysmal... @kavir1698 when you added the code based on this SQLdata base you claimed it was "fast enough" but that is not what I see here. Moving an agent shouldn't take more than nanoseconds, so all the time is spent in updating the database. @kavir1698 I think you should do this new compartment space once I completely replace `GridSpace` with my `ArraySpace`.

Yes the continuous space was always slower than grid spaces. The database shows its advantage in large models. That is what we saw from previous benchmarks. We will implement the compartment space, but we should also keep the current continuous space because of the limitation of the compartment space we had discussed.

**Comment by Datseris** [comment]
> We will implement the compartment space, but we should also keep the current continuous space because of the limitation of the compartment space we had discussed.

The compartment space has **no limitations** versus the current space, but I think this misunderstanding exists because I haven't made clear my idea on the compartment space. I'll do so in a clear description in a new issue.

**Comment by Libbum** [comment]
I'll post a new output once it's done, but will be AFK for the afternoon&mdash;teaching on-site today.

**Comment by Datseris** [comment]
Thanks Tim. This is good to go from my eyes.

File: Project.toml [hunk 2]
```diff
@@ -8,6 +8,7 @@ DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
 Distributed = "8ba89e20-285c-5b6f-9357-94700520ee1b"
 LightGraphs = "093fc24a-ae57-5d10-9952-331d41423f4d"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+PkgBenchmark = "32113eaa-f34f-5b0d-bd6c-c81e245fc73d"
```
Comment (Datseris) [inline Project.toml:11]: Wait, this shouldn't be here.

Comment (Libbum) [reply Project.toml:11]: Ah. Yeah, I'll dump that.

**Comment by Libbum** [comment]
```julia
Benchmarkresults:
    Package: /mnt/turtle/repos/Agents.jl
    Date: 18 Sep 2020 - 12:04
    Package commit: 8fae9a
    Julia commit: 96786e
    BenchmarkGroup:
        5-element BenchmarkTools.BenchmarkGroup:
          tags: ["Agents"]
          "space_creation" => 3-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graph", "grid", "continuous"]
                  "grid" => Trial(62.803 ms)
                  "graph" => Trial(2.187 ms)
                  "continuous" => Trial(120.432 μs)
          "model" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["initialise", "initialise_union"]
                  "initialise_union" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["graph", "grid", "continuous"]
                          "grid" => Trial(5.986 μs)
                          "graph" => Trial(5.616 μs)
                          "continuous" => Trial(6.203 μs)
                  "initialise" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["graph", "grid", "continuous"]
                          "grid" => Trial(1.040 μs)
                          "graph" => Trial(1.031 μs)
                          "continuous" => Trial(1.503 μs)
          "grid" => 6-element BenchmarkTools.BenchmarkGroup:
                  tags: ["add", "add_union", "move", "neighbors", "node"]
                  "add" => 7-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single", "create", "create_pos", "create_single", "create_fill"]
                          "create_single" => Trial(10.688 ms)
                          "create" => Trial(1.912 μs)
                          "agent_single" => Trial(10.664 ms)
                          "create_fill" => Trial(107.633 μs)
                          "create_pos" => Trial(222.008 μs)
                          "agent" => Trial(198.709 ns)
                          "agent_pos" => Trial(24.554 ns)
                  "add_union" => 4-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single", "agent_fill"]
                          "agent_single" => Trial(10.649 ms)
                          "agent_fill" => Trial(116.285 μs)
                          "agent" => Trial(213.358 ns)
                          "agent_pos" => Trial(36.832 ns)
                  "collect" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["init_agent", "store_agent"]
                          "store_agent" => Trial(1.340 ms)
                          "init_agent" => Trial(1.637 μs)
                  "neighbors" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["space_pos", "space_agent", "space_pos_iterate", "space_agent_iterate", "node_pos", "node_agent"]
                          "space_agent_iterate" => Trial(11.260 ms)
                          "node_pos" => Trial(159.021 ns)
                          "node_agent" => Trial(173.000 ns)
                          "space_pos" => Trial(40.317 μs)
                          "space_pos_iterate" => Trial(5.957 ms)
                          "space_agent" => Trial(51.844 μs)
                  "node" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["contents", "agents"]
                          "agents" => Trial(84.000 ns)
                          "contents" => Trial(2.308 ns)
                  "move" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["random", "pos", "single"]
                          "random" => Trial(62.095 ns)
                          "pos" => Trial(28.052 ns)
                          "single" => Trial(2.404 μs)
          "graph" => 6-element BenchmarkTools.BenchmarkGroup:
                  tags: ["add", "add_union", "move", "neighbors", "node"]
                  "add" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single", "create", "create_pos", "create_single"]
                          "create_single" => Trial(294.761 μs)
                          "create" => Trial(47.179 μs)
                          "agent_single" => Trial(343.398 μs)
                          "create_pos" => Trial(156.747 μs)
                          "agent" => Trial(93.185 ns)
                          "agent_pos" => Trial(22.134 ns)
                  "add_union" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "agent_single"]
                          "agent_single" => Trial(284.759 μs)
                          "agent" => Trial(102.060 ns)
                          "agent_pos" => Trial(35.722 ns)
                  "collect" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["init_agent", "store_agent"]
                          "store_agent" => Trial(37.849 μs)
                          "init_agent" => Trial(1.670 μs)
                  "neighbors" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["space_pos", "space_agent", "space_pos_iterate", "space_agent_iterate", "node_pos", "node_agent"]
                          "space_agent_iterate" => Trial(29.719 μs)
                          "node_pos" => Trial(19.440 ns)
                          "node_agent" => Trial(21.100 ns)
                          "space_pos" => Trial(1.444 μs)
                          "space_pos_iterate" => Trial(29.823 μs)
                          "space_agent" => Trial(1.612 μs)
                  "node" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["contents", "agents"]
                          "agents" => Trial(69.200 ns)
                          "contents" => Trial(1.732 ns)
                  "move" => 3-element BenchmarkTools.BenchmarkGroup:
                          tags: ["random", "pos", "single"]
                          "random" => Trial(36.069 ns)
                          "pos" => Trial(22.292 ns)
                          "single" => Trial(100.000 ns)
          "continuous" => 5-element BenchmarkTools.BenchmarkGroup:
                  tags: ["add", "add_union", "move", "neighbors"]
                  "add" => 4-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos", "create", "create_pos"]
                          "create" => Trial(15.358 μs)
                          "create_pos" => Trial(232.209 μs)
                          "agent" => Trial(7.759 μs)
                          "agent_pos" => Trial(5.643 μs)
                  "add_union" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["agent", "agent_pos"]
                          "agent" => Trial(8.226 μs)
                          "agent_pos" => Trial(6.058 μs)
                  "collect" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["init_agent", "store_agent"]
                          "store_agent" => Trial(7.555 ms)
                          "init_agent" => Trial(1.746 μs)
                  "neighbors" => 6-element BenchmarkTools.BenchmarkGroup:
                          tags: ["space_pos", "space_agent", "space_pos_iterate", "space_agent_iterate", "nearest", "interacting"]
                          "space_agent_iterate" => Trial(15.566 ms)
                          "nearest" => Trial(6.192 ms)
                          "space_pos" => Trial(10.070 ms)
                          "space_pos_iterate" => Trial(15.172 ms)
                          "interacting" => Trial(142.081 s)
                          "space_agent" => Trial(5.969 ms)
                  "move" => 2-element BenchmarkTools.BenchmarkGroup:
                          tags: ["update", "vel"]
                          "update" => Trial(2.082 μs)
                          "vel" => Trial(2.128 μs)
```

**Pull request merged**